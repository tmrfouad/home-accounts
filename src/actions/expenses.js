import database from '../firebase/firebase';

// Add_EXPENSE
export const addExpense = expense => ({
  type: 'Add_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      account = {
        id: '',
        name: ''
      },
      description = '',
      notes = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = {
      account,
      description,
      notes,
      amount,
      createdAt
    };

    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(addExpense({ id: ref.key, ...expense }));
      })
      .catch(() => {});
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

// EDIT_EXPENSE
export const editExpense = ({ id } = {}, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = ({ id } = {}, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense({ id }, updates));
      });
  };
};

// SET_EXPENSES
export const setExpenses = expenses => ({ type: 'SET_EXPENSES', expenses });

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value', snap => {
      const expenses = [];
      snap.forEach(childSnap => {
        expenses.push({
          id: childSnap.key,
          ...childSnap.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  };
};
