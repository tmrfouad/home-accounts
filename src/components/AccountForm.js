import React from 'react';
import { connect } from 'react-redux';
import {
  startAddAccount,
  startEditAccount,
  startRemoveAccount
} from '../actions/accounts';
import ConfirmModal from './ConfirmModal';

export class AccountForm extends React.Component {
  constructor(props) {
    super(props);
    const account = this.props.account;
    this.state = {
      name: account ? account.name : '',
      error: '',
      mode: account ? 'edit' : 'add',
      isModalOpen: false
    };
  }

  onNameChange = e => {
    this.setState({ name: e.target.value });
  };

  onFormSubmit = e => {
    e.preventDefault();
    if (!this.state.name) {
      this.setState({ error: 'Please enter a name!' });
    } else {
      if (this.state.mode === 'edit') {
        this.props.startEditAccount(this.props.account.id, {
          name: this.state.name
        });
      } else {
        if (this.props.accounts.find(acc => acc.name === this.state.name)) {
          this.setState({ error: 'This account already exists!' });
        } else {
          this.props.startAddAccount({
            name: this.state.name
          });
        }
      }
      this.props.history.push('/accounts');
    }
  };

  removeAccount = () => {
    this.props.startRemoveAccount(this.props.account.id);
    this.props.history.push('/accounts');
  };

  openRemoveItemDialog = e => {
    e.preventDefault();
    this.setState(() => ({ isModalOpen: true }));
  };

  closeRemoveItemDialog = () => {
    this.setState(() => ({ isModalOpen: false }));
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">
              {this.state.mode === 'edit' ? 'Edit' : 'Add'} Account
            </h2>
          </div>
        </div>
        <form className="form content-container" onSubmit={this.onFormSubmit}>
          {this.state.error && (
            <div className="form__error">{this.state.error}</div>
          )}
          <input
            type="text"
            className="text-input"
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <div>
            <button className="button form__action">Save Account</button>
            <button
              className="button form__action button--secondary"
              onClick={this.openRemoveItemDialog}
            >
              Remove Account
            </button>
          </div>
        </form>
        <ConfirmModal
          id="confirmModal"
          messageTitle="Remove Account!"
          messageBody="Are you sure you want to remove this item?"
          isOpen={this.state.isModalOpen}
          onModalClose={this.closeRemoveItemDialog}
          onModalOk={this.removeAccount}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  account: state.accounts.find(t => t.id === props.match.params.id),
  accounts: state.accounts
});

const mapDispatchToProps = dispatch => ({
  startAddAccount: account => dispatch(startAddAccount(account)),
  startEditAccount: (id, account) =>
    dispatch(startEditAccount({ id }, account)),
  startRemoveAccount: id => dispatch(startRemoveAccount({ id }))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountForm);
