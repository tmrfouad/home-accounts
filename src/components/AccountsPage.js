import React from 'react';
import AccountList from './AccountList';

const AccountsPage = () => (
  <div>
    <AccountList />
  </div>
);

export default AccountsPage;
