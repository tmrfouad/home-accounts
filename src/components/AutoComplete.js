import React from 'react';

export default class AutoComplete extends React.Component {
  constructor(props) {
    super(props);

    if (
      !props.id ||
      !props.source ||
      !props.displayField ||
      !props.valueField ||
      !props.onChange
    ) {
      throw new Error(
        'You must supply the id, src, displayField, valueField and onChange properties for the AutoComplete component.'
      );
    }
  }

  updateInputs() {
    const selectedItem =
      this.props.source.find(
        itm => itm[this.props.valueField] + '' === this.props.value + ''
      ) || {};

    const textInput = document.getElementById(this.props.id + '-text');
    const hiddenInput = document.getElementById(this.props.id + '-hidden');

    textInput.value = selectedItem[this.props.displayField] || '';
    hiddenInput.value = selectedItem[this.props.valueField] || '';
  }

  componentDidUpdate() {
    this.updateInputs();
  }

  componentDidMount() {
    this.updateInputs();
    this.autocomplete();
  }

  autocomplete() {
    const self = this;
    const inp = document.getElementById(self.props.id + '-text');
    const src = self.props.source;
    const displayField = self.props.displayField;
    const valueField = self.props.valueField;
    const onChange = self.props.onChange;

    let currentFocus;

    inp.addEventListener('click', function() {
      showList(true);
    });

    inp.addEventListener('blur', function(e) {
      if (e.target.closest('.autocomplete')) {
        return;
      }
      let inputValue = inp.value;
      const foundObj = src.find(o => o[displayField] + '' === inputValue + '');
      if (foundObj) {
        const value = foundObj[valueField];
        onChange(value);
        closeAllLists();
      } else {
        if (isListOpen()) {
          const firstListItem = inp.parentNode.querySelector(
            '.autocomplete-active'
          );
          if (firstListItem) {
            const value = firstListItem.getElementsByTagName('input')[0].value;
            const firstObj = src.find(
              itm => itm[valueField] + '' === value + ''
            );
            onChange(value);
            closeAllLists();
          }
        }
      }
    });

    /*execute a function when someone writes in the text field:*/
    inp.addEventListener('input', function(e) {
      showList();
    });
    /*execute a function presses a key on the keyboard:*/
    inp.addEventListener('keydown', function(e) {
      let x = document.getElementById(this.id + 'autocomplete-list');
      if (x) x = x.getElementsByTagName('div');
      if (e.keyCode == 40) {
        if (isListOpen()) {
          /*If the arrow DOWN key is pressed,
          increase the currentFocus variable:*/
          currentFocus++;
          /*and and make the current item more visible:*/
          addActive(x);
        } else {
          showList(true);
        }
      } else if (e.keyCode == 38) {
        //up
        /*If the arrow UP key is pressed,
        decrease the currentFocus variable:*/
        currentFocus--;
        /*and and make the current item more visible:*/
        addActive(x);
      } else if (e.keyCode == 13) {
        /*If the ENTER key is pressed, prevent the form from being submitted,*/
        e.preventDefault();
        if (currentFocus > -1) {
          /*and simulate a click on the "active" item:*/
          if (x) x[currentFocus].click();
        }
      } else if (e.keyCode == 27) {
        closeAllLists();
      }
    });
    function showList(showAll) {
      if (showAll && isListOpen()) {
        closeAllLists();
        return;
      }

      let a,
        b,
        i,
        inputValue = inp.value;

      /*close any already open lists of autocompleted values*/
      closeAllLists();

      if (!showAll && !inputValue) {
        return false;
      }
      currentFocus = -1;
      /*create a DIV element that will contain the items (values):*/
      a = document.createElement('DIV');
      a.setAttribute('id', inp.id + 'autocomplete-list');
      a.setAttribute('class', 'autocomplete-items');
      /*append the DIV element as a child of the autocomplete container:*/
      inp.parentNode.appendChild(a);
      /*for each item in the array...*/
      for (i = 0; i < src.length; i++) {
        /*check if the item starts with the same letters as the text field value:*/
        const srcObj = src[i];
        const srcValue = srcObj[valueField];
        const srcDisplay = srcObj[displayField];
        if (
          srcDisplay.toUpperCase().includes(inputValue.toUpperCase()) ||
          showAll
        ) {
          /*create a DIV element for each matching element:*/
          b = document.createElement('DIV');
          if (showAll) {
            if (srcDisplay.toUpperCase() === inputValue.toUpperCase()) {
              b.innerHTML = '<strong>' + srcDisplay + '</strong>';
            } else {
              b.innerHTML = srcDisplay;
            }
          } else {
            /*make the matching letters bold:*/
            const valIndex = srcDisplay
              .toUpperCase()
              .indexOf(inputValue.toUpperCase());
            b.innerHTML = srcDisplay.substr(0, valIndex);
            b.innerHTML +=
              '<strong>' +
              srcDisplay.substr(valIndex, inputValue.length) +
              '</strong>';
            b.innerHTML += srcDisplay.substr(valIndex + inputValue.length);
          }

          /*insert a input field that will hold the current array item's value:*/
          b.innerHTML += "<input type='hidden' value='" + srcValue + "'>";
          /*execute a function when someone clicks on the item value (DIV element):*/
          b.addEventListener('click', function(e) {
            /*insert the value for the autocomplete text field:*/
            const value = this.getElementsByTagName('input')[0].value;
            const selectedItem = self.props.source.find(
              itm => itm[valueField] + '' === value + ''
            );
            const display = selectedItem
              ? selectedItem[self.props.displayField]
              : '';

            // inp.value = display;
            // self.state.value[displayField] = display;

            // valInp.value = value;
            /*close the list of autocompleted values,
            (or any other open lists of autocompleted values:*/
            closeAllLists();
            onChange(value);
          });
          a.appendChild(b);
        }
      }
    }
    function addActive(x) {
      /*a function to classify an item as "active":*/
      if (!x) return false;
      /*start by removing the "active" class on all items:*/
      removeActive(x);
      if (currentFocus >= x.length) currentFocus = 0;
      if (currentFocus < 0) currentFocus = x.length - 1;
      /*add class "autocomplete-active":*/
      x[currentFocus].classList.add('autocomplete-active');
    }
    function removeActive(x) {
      /*a function to remove the "active" class from all autocomplete items:*/
      for (var i = 0; i < x.length; i++) {
        x[i].classList.remove('autocomplete-active');
      }
    }
    function closeAllLists(elmnt) {
      /*close all autocomplete lists in the document,
      except the one passed as an argument:*/
      let x = inp.parentNode.getElementsByClassName('autocomplete-items');
      for (var i = 0; i < x.length; i++) {
        if (elmnt != x[i] && elmnt != inp) {
          x[i].parentNode.removeChild(x[i]);
        }
      }
    }
    function isListOpen() {
      return !!inp.parentNode.querySelector('.autocomplete-items');
    }
    /*execute a function when someone clicks in the document:*/
    document.addEventListener('click', function(e) {
      closeAllLists(e.target);
    });
  }

  // <div className="backline" />

  render() {
    return (
      <div className="autocomplete">
        <input
          id={this.props.id + '-text'}
          type="text"
          className={this.props.className}
          placeholder={this.props.placeholder}
          autoComplete="off"
        />
        <input id={this.props.id + '-hidden'} type="hidden" />
      </div>
    );
  }
}
