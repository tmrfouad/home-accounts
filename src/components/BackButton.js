import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft } from '@fortawesome/free-solid-svg-icons';
import { history } from '../routers/AppRouter';

export const BackButton = () => (
  <button
    className="button button--link--dark"
    href="javascript:void(0);"
    onClick={() => history.goBack()}
  >
    <FontAwesomeIcon icon={faChevronLeft} /> Back
  </button>
);
