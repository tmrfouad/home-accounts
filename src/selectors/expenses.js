import moment from 'moment';

// Get visible expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      return (
        (!text ||
          expense.description.toLowerCase().includes(text.toLowerCase())) &&
        (!startDate || moment(expense.createdAt).isSameOrAfter(startDate)) &&
        (!endDate || moment(expense.createdAt).isSameOrBefore(endDate))
      );
    })
    .sort((a, b) => {
      if (!sortBy) {
        return 0;
      }

      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }

      return 0;
    });
};
