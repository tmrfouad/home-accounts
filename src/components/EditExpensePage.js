import React from 'react';
import ExpenseForm from './ExpenseForm';
import ConfirmModal from './ConfirmModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import { connect } from 'react-redux';

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isModalOpen: false
    };
  }

  onSubmit = exp => {
    this.props.startEditExpense(this.props.expense.id, exp);
    this.props.history.push('/');
  };

  onRemoveExpense = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  openDialog = () => {
    this.setState(() => ({ isModalOpen: true }));
  };

  closeDialog = () => {
    this.setState(() => ({ isModalOpen: false }));
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Edit Expense</h2>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            id="expenseForm"
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button
            className="button button--secondary"
            onClick={this.openDialog}
          >
            Remove Expense
          </button>
        </div>
        <ConfirmModal
          id="confirmModal"
          messageTitle="Remove Expense!"
          messageBody="Are you sure you want to remove this item?"
          isOpen={this.state.isModalOpen}
          onModalClose={this.closeDialog}
          onModalOk={this.onRemoveExpense}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, exp) => {
    dispatch(startEditExpense({ id }, exp));
  },
  startRemoveExpense: id => {
    dispatch(startRemoveExpense({ id }));
  }
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
