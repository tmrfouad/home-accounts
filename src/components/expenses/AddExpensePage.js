import React from 'react';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../../actions/expenses';
import { connect } from 'react-redux';
import { BackButton } from '../BackButton';

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div
            className={
              this.props.styles.collapsed
                ? 'content-container content-container--collapsed'
                : 'content-container'
            }
          >
            <BackButton />
            <h2 className="page-header__title">Add Expense</h2>
          </div>
        </div>
        <div
          className={
            this.props.styles.collapsed
              ? 'content-container content-container--collapsed'
              : 'content-container'
          }
        >
          <ExpenseForm id="expenseForm" onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  styles: state.styles
});

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddExpensePage);
