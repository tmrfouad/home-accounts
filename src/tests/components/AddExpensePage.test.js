import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

let startAddExpense, history, wrapper;
beforeEach(() => {
  startAddExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <AddExpensePage history={history} startAddExpense={startAddExpense} />
  );
});

test('should rnder AddExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('#expenseForm').prop('onSubmit')(expenses[0]);
  expect(startAddExpense).toHaveBeenLastCalledWith(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
});
