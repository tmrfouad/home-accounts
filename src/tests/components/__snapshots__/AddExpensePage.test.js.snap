// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should rnder AddExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        Add Expense
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <Connect(ExpenseForm)
      id="expenseForm"
      onSubmit={[Function]}
    />
  </div>
</div>
`;
